### PR TITLE
Add tl2cgen feedstock

### PR DESCRIPTION
**Pull request by hcho3**
Parts of the Treelite package are being migrated into a brand new package called TL2cgen (TreeLite 2 C Generator). I am submitting this PR to add TL2cgen to Conda-forge.

From [TL2cgen doc](https://tl2cgen.readthedocs.io/en/latest/#how-are-treelite-and-tl2cgen-related):
> **How are Treelite and TL2cgen related?**
> Starting from 4.0 version, Treelite no longer supports compiling tree models into C code. That part of Treelite has been migrated to TL2cgen.
> * **Treelite** (starting from 4.0) is now a small libary that enables other C++ applications to exchange and store decision trees on the disk as well as the network. By using Treelite, application writers can support many kinds of tree models with minimal code duplication and high level of correctness. Trees are stored using an efficient binary format.
> * **TL2cgen** is a model compiler that converts tree models into C code. It can convert any tree models that can be stored using the Treelite format. TL2cgen is one of multiple applications that uses Treelite as a library.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

File: recipes/tl2cgen/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    - fmt
+    - rapidjson
+    - llvm-openmp  # [osx]
+    - libgomp  # [linux]
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+  host:
+    - python
+    - pip >=22.1
+    - treelite>=3.4,<4
```
Comment (hcho3) [inline recipes/tl2cgen/meta.yaml:30]: I will remove the constraint `<4` once Treelite 4.0 is released. I plan to make many breaking changes to Treelite 4.0.

File: recipes/tl2cgen/0001-windows-libpath-fix.patch
Comment (hcho3) [inline recipes/tl2cgen/0001-windows-libpath-fix.patch:1]: I plan to upstream this patch for the next release (0.2.0).

**Comment by hcho3** [comment]
@conda-forge-admin, please lint

**Comment by hcho3** [comment]
@conda-forge/staged-recipes, please review

File: recipes/tl2cgen/build.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+set -x
+set -e
+
+mkdir build
+cd build
+cmake ${CMAKE_ARGS} .. -DCMAKE_INSTALL_PREFIX="$PREFIX" -DCMAKE_INSTALL_LIBDIR="lib" -DCMAKE_CXX_FLAGS="-D_LIBCPP_DISABLE_AVAILABILITY" -DCMAKE_BUILD_TYPE=Release -GNinja
+ninja install   # Install libtl2cgen.so into Conda env
```
Comment (ocefpaf) [inline recipes/tl2cgen/build.sh:9]: Is there a make check or ctest to run here?

File: recipes/tl2cgen/0001-windows-libpath-fix.patch [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+-            dll_path.append(curr_path.joinpath("../../windows/Release/"))
+-            dll_path.append(curr_path.joinpath("./windows/Release/"))
++        # On Windows, Conda may install libs in different paths
++        sys_prefix = pathlib.Path(sys.prefix)
++        dll_path.extend(
++            [
++                sys_prefix / "bin",
++                sys_prefix / "Library",
++                sys_prefix / "Library" / "bin",
++                sys_prefix / "Library" / "lib",
```
[the remark this answers is not in the document]
Comment (hcho3) [reply recipes/tl2cgen/0001-windows-libpath-fix.patch:85]: Sure, I'll create another release upstream to incorporate this patch.

**Comment by hcho3** [comment]
@ocefpaf  For some reason, the C++ test is getting stuck on Windows. I could not reproduce it on my end. (It runs very quickly on my Windows machine.) Is it okay if we take out `ctest -V` from `bld.bat`?

**Comment by ocefpaf** [comment]
> Is it okay if we take out `ctest -V` from `bld.bat`?

Sure. Let's give it a try. If nothing works we can skip it for now and try again when a new version is released.

**Comment by hcho3** [comment]
Thank you!

**Pull request merged**